Unable to build with glibc-2.34

2.10.0-beta1-317-gc9787d572

> cmake . -DCMAKE_BUILD_TYPE=Debug -DENABLE_BACKTRACE=ON -DENABLE_DIST=ON -DENABLE_FEEDBACK_DAEMON=OFF -DENABLE_BUNDLED_LIBCURL=OFF -DENABLE_FEEDBACK_DAEMON=OFF && make -j

```c
[ 81%] Building CXX object test/unit/CMakeFiles/guard.test.dir/guard.cc.o
In file included from /usr/include/signal.h:328,
                 from /home/alex/p/tarantool-meta/r/t-5/third_party/libev/ev.h:167,
                 from /home/alex/p/tarantool-meta/r/t-5/third_party/tarantool_ev.h:54,
                 from /home/alex/p/tarantool-meta/r/t-5/src/lib/core/say.h:39,
                 from /home/alex/p/tarantool-meta/r/t-5/src/lib/core/diag.h:41,
                 from /home/alex/p/tarantool-meta/r/t-5/src/lib/core/exception.h:38,
                 from /home/alex/p/tarantool-meta/r/t-5/src/lib/small/include/small/region.h:277,
                 from /home/alex/p/tarantool-meta/r/t-5/src/lib/core/memory.h:33,
                 from /home/alex/p/tarantool-meta/r/t-5/test/unit/guard.cc:1:
/home/alex/p/tarantool-meta/r/t-5/test/unit/guard.cc:31:23: error: size of array ‘stack_buf’ is not an integral constant-expression
   31 | static char stack_buf[SIGSTKSZ];
      |                       ^~~~~~~~
cc1plus: note: unrecognized command-line option ‘-Wno-gnu-alignof-expression’ may have been intended to silence earlier diagnostics
```

--- a/changelogs/unreleased/gh-6686-build-with-glibc-2-34.md
+++ b/changelogs/unreleased/gh-6686-build-with-glibc-2-34.md
@@ -1,0 +1,4 @@
+## bugfix/build
+
+* Fix build errors with glibc-2.34 (gh-6686).
+* Change size of alt. signal stack for ASAN needs.

--- a/test/unit/guard.cc
+++ b/test/unit/guard.cc
@@ -28,16 +28,24 @@ stack_break_f(char *ptr)
 	return sum;
 }
 
-static char stack_buf[SIGSTKSZ];
-
 static int
 main_f(va_list ap)
 {
 	stack_t stack;
-	stack.ss_sp = stack_buf;
-	stack.ss_size = SIGSTKSZ;
 	stack.ss_flags = 0;
-	sigaltstack(&stack, NULL);
+	/*
+	 * It is said that SIGSTKSZ is not enough for one of llvm sanitizers
+	 * (probably asan, because this test fails with segmentation fault if
+	 * we use SIGSTKSZ as alternative signal stack size when we use it).
+	 * https://github.com/llvm/llvm-project/blame/699231ab3c7dd8f028d868b103481fa901f3c721/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp#L169
+	 */
+	stack.ss_size = 4 * SIGSTKSZ;
+	stack.ss_sp = xmalloc(stack.ss_size);
+	if (sigaltstack(&stack, NULL) < 0) {
+		free(stack.ss_sp);
+		perror("sigaltstack");
+		exit(EXIT_FAILURE);
+	}
 	struct sigaction sa;
 	sa.sa_handler = sigsegf_handler;
 	sigemptyset(&sa.sa_mask);
@@ -48,6 +56,7 @@ main_f(va_list ap)
 	int res = stack_break_f((char *)&stack);
 
 	ev_break(loop(), EVBREAK_ALL);
+	free(stack.ss_sp);
 	return res;
 }
 

